order/cache.go

```diff
@@ -0,0 +1,110 @@
+package order
+
+import (
+	"bytes"
+	"encoding/gob"
+	"io/ioutil"
+	"log"
+	"time"
+
+	"github.com/allegro/bigcache"
+	"github.com/foomo/shop/order"
+	"github.com/foomo/shop/state"
+	"github.com/foomo/shop/version"
+)
+
+var cache *bigcache.BigCache
+
+func init() {
+
+	log.Println("init order cache")
+
+	gob.Register(time.Time{})
+	gob.Register(order.Position{})
+	gob.Register(order.CustomerData{})
+	gob.Register(order.Processing{})
+	gob.Register(state.State{})
+	gob.Register(order.Flags{})
+	gob.Register(version.Version{})
+	gob.Register(order.Order{})
```

Comment:
If we're having custom data in the order (globus-wise) this might break if the corresponding JSON entry has some non-registered fields (custom fields). Perhaps there would be a better, more accomodating format?

Since if we break for not serializing a lets say map[string]interface{} this will break the order service...

Reply:
Good point. 

I am not too familiar with the data carried by an Order instance,
a glimpse on the type definition reveals a lot of nested structures.

All of these and all their fields would need to be registered when using gob.

```go
type Order struct {
	ID                         string           `json:"id"`
	State                      OrderState       `json:"state"` //foomo_shop_state.State // z.B. In Bearbeitung, versendet
	StateLastModifiedTimestamp int64            `json:"stateLastModifiedTimestamp"`
	IsApprovedForCheckout      bool             `json:"isApprovedForCheckout"` // true if all positions are available / valid and the user could continue with checkout
	Positions                  []*OrderPosition `json:"positions"`
	LanguageCode               LanguageCode     `json:"languageCode"`
	Notifications              []*Notification  `json:"notifications"`
	Summary                    *OrderSummary    `json:"summary"`
	PaymentInfos               []*PaymentInfo   `json:"paymentInfos"`
	HasExternalPayment         bool             `json:"hasExternalPayment"` // true, if payment was done in-store (e.g. cash). => payment at cashdesk
	RefundInfos                []*RefundInfo    `json:"refundInfos"`
	CustomerData               *CustomerData    `json:"customerData"`
	LastModified               string           `json:"lastModified"`
	// Fields relevant for overview of confirmed and completed orders
	DateOfOrder           string            `json:"dateOfOrder"`
	DeliveryInfo          *DeliveryInfo     `json:"deliveryInfo"`      // For now, there is only one DeliverInfo for the whole order (It's a slice on foomo shop order)
	OrderPartialTypes     []string          `json:"orderPartialTypes"` // a string array containing all possible types which are the possible keys for Positions map.
	BillingDownloadLink   string            `json:"billingDownloadLink"`
	CartTrackingInfo      *CartTrackingInfo `json:"cartTrackingInfo"`
	Channel               SalesChannel      `json:"salesChannel"`
	Salesperson           *Salesperson      `json:"salesperson"`
	OmniChannel           *OmniChannel      `json:"omnichannel"`
	Processing            *Processing       `json:"processing"`
	TotalQuantity         float64           `json:"totalQuantity"`
	TotalCanceledQuantity float64           `json:"totalCanceledQuantity"`
	TotalReturnedQuantity float64           `json:"totalReturnedQuantity"`
	RetailDepartment      string            `json:"retailDepartment"`
}
```

I'd suggest messagepack, since it does not have the requirement of pre-registering the structures,
has a small memory footprint and fast serialisation.

However, maybe it would be sufficient to implement a sync.Map of orderIDs to order instances and a collection routine, instead of using bigcache? This way there would be no need for serialisation at all.

Reply:
I would also favour the sync.Map approach to prevent running into serialisation issues.

Reply:
BigCache keeps entries on heap but omits GC for them. To achieve that operations on bytes arrays take place, therefore entries (de)serialization in front of the cache will be needed in most use cases.
But switching to messagepack might be an option. I also wasn't happy with gob registration here …

Reply:
So, we can have https://github.com/ugorji/go/tree/master/codec which allows message pack encoding, + on the fly code generation for objects to do less allocation.

Reply:
sync.Map is not suggested to be used on high frequency threaded access. It's safe, but slooow since it is using locks.
After reading that article https://blog.dgraph.io/post/caching-in-go/ I decided to give bigcache a try.

Reply:
@loeffert that's true, although the comparisons are in millions of requests/second and we have like 15 requests per second for that particular cache, which is kind of an overkill.

I mean like I said, we could use a helicopter to go to the store in the local corner 😄 

But I'm all for anything that will remove the big S in Sirupsen 😸